unixfs/src/dir/builder/iter.rs

```diff
@@ -0,0 +1,416 @@
+                    // Cids in the btreemap
+                    let mut collected = self.persisted_cids.remove(&Some(id)).unwrap_or_default();
+
+                    // FIXME: leaves could be drained and reused
+                    collected.extend(leaves);
+
+                    if !self.wrap_with_directory && parent_id.is_none() {
+                        // we aren't supposed to wrap_with_directory, and we are now looking at the
+                        // possibly to be generated root directory.
+
+                        assert_eq!(
+                            collected.len(),
+                            1,
+                            "should not have gone this far with multiple added roots"
+                        );
+
+                        return None;
+                    }
+
+                    let buffer = &mut self.block_buffer;
+
+                    let leaf = match Self::render_directory(&collected, buffer) {
+                        Ok(leaf) => leaf,
+                        Err(e) => return Some(Err(e)),
+                    };
+
+                    self.cid = Some(leaf.link.clone());
+                    self.total_size = leaf.total_size;
+
+                    // this reuse strategy is probably good enough
+                    collected.clear();
+
+                    if let Some(name) = name {
+                        // name is none only for the wrap_with_directory, which cannot really be
+                        // propagated up but still the parent_id is allowed to be None
+                        let previous = self
+                            .persisted_cids
+                            .entry(parent_id)
+                            .or_insert(collected)
+                            .insert(name, leaf);
+
+                        assert!(previous.is_none());
+                    }
+
+                    return Some(Ok(TreeNode {
+                        path: self.full_path.as_str(),
+                        cid: self.cid.as_ref().unwrap(),
+                        total_size: self.total_size,
+                        block: &self.block_buffer,
+                    }));
+                }
+            }
+        }
+        None
+    }
+}
+
+impl<'a> Iterator for PostOrderIterator<'a> {
+    type Item = Result<OwnedTreeNode, TreeConstructionFailed>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.next_borrowed()
+            .map(|res| res.map(TreeNode::into_owned))
+    }
+}
+
+/// Borrowed representation of a node in the tree.
+pub struct TreeNode<'a> {
+    /// Full path to the node.
+    pub path: &'a str,
+    /// The Cid of the document.
+    pub cid: &'a Cid,
+    /// Cumulative total size of the subtree in bytes.
+    pub total_size: u64,
+    /// Raw dag-pb document.
+    pub block: &'a [u8],
+}
+
+impl TreeNode<'_> {
+    /// Convert to an owned and detached representation.
+    pub fn into_owned(self) -> OwnedTreeNode {
```

Comment:
nit: there's always https://doc.rust-lang.org/std/borrow/trait.ToOwned.html

Reply:
there's also https://github.com/koivunej/derive-into-owned but I'll check if ToOwned can be used here.

Reply:
I don't think ToOwned can be used here at all since we don't have GAT or does std::borrow::Borrow have a lifetime. Please confirm I didn't miss anything?

Reply:
No worries; I expected that it could potentially be a pain to implement, hence the "nit" 😛